versioned_hdf5/backend.py

```diff
@@ -68,17 +74,117 @@ def initialize(f):
+        fletcher32: None,  # True if present; False if missing
+    }
+
+    Note: you should not assume that the above is the complete content
+    of _dataset._filters. There may be custom/unknown filters present.
+    """
+
+    # See matching class wrappers.FiltersMixin
+    compression: Any | None | Default = DEFAULT
+    compression_opts: Any | None | Default = DEFAULT
+    scaleoffset: int | None | Default = DEFAULT
+    shuffle: bool | Default = DEFAULT
+    fletcher32: bool | Default = DEFAULT
+
+    def as_kwargs(self) -> dict[str, Any]:
+        """Convert to kwargs for create_dataset."""
+        return {k: v for k, v in self.__dict__.items() if v is not DEFAULT}
+
+    @staticmethod
+    def from_dataset(ds: Dataset | FiltersMixin) -> Filters:
+        """Reverse engineer create_dataset kwargs for the filters from an existing
+        h5py.Dataset.
+        """
+        compression = ds.compression
+        compression_opts = ds.compression_opts
+        # FIXME This should be fixable upstream. h5py would need to expose a hook for
+        # hdf5plugin / pytables to let them declare that a filter ID is a
+        # compression filter.
+
+        if compression is None and isinstance(ds, Dataset):
+            # From hdf5plugin._filters. Can't just try-import hdf5plugin because the same
+            # filter IDs are also defined by pytables.
+            CUSTOM_COMPRESSION_FILTERS = {
+                32001,  # Blosc
+                32026,  # Blosc2
+                307,  # Bzip2
+                32004,  # LZ4
+                32008,  # Bitshuffle
+                32013,  # ZFP
+                32015,  # Zstandard
+                32017,  # SZ
+                32024,  # SZ3
+                32018,  # FCIDECOMP
+                32028,  # SPERR
+            }
+
+            for filter_id in CUSTOM_COMPRESSION_FILTERS:
+                try:
+                    compression_opts = ds._filters[str(filter_id)]
+                    compression = filter_id
+                    break
+                except KeyError:
+                    pass
+
+            if compression is None:
+                # If we're using a bespoke compression, there's no way of knowing
+                # whether an unknown filter is a valid compression or some other
+                # kind of filter, so we issue a warning about assuming that it is
+                # the dataset's compression.
+                for k, v in ds._filters.items():
+                    try:
+                        k = int(k)
+                    except ValueError:
+                        continue
+                    compression, compression_opts = k, v
+                    logging.warning(
+                        "No default compression detected in this dataset. "
+                        f"Guessed {compression=} {compression_opts=} "
+                        f"from {ds._filters=}."
+                    )
+                    break
```

Comment:
This code path is untested. I'm unsure how to write a test for it.
The alternative was not to whitelist blosc etc. and just run this path. However, that carried the risk of encountering a custom non-compression filter plus blosc, which would mangle things up.